### PR TITLE
Review `load_data`

### PR DESCRIPTION
**Pull request by JoeZiminski**
This PR is to review the `load_data` code from spikewrap. 

The PR system is a little strange but I think should (hopefully) work well. The aim of this system is avoid having  to review the entire codebase at once, or review since-reverted commits without changing the `dev` / `main` brain git history. `reviewed_code` was created by branching `dev` and removing all commits. `rev_load_data` was created by branching `dev` and removing all of the code unrelated to loading data. Now merging `rev_load_data > reviewed code` should isolate these commits of interest. For new features, I will just add them from `dev` to here. Once the main bulk of the code is reviewed we can revert to reviewing in the normal way. This is a very weird system :'D but it should work out ok! and allow tailored commits to be reviewed.

This PR includes all the general package-related files (e.g. `pyproject.toml`) as well as:

`examples/load_data.py` : an example of how to use the code to load data into `spikewrap`.
`pipeline/load_data.py` : wrapper code to load the raw data into a class data structure. Under the hood the data is loaded with spikeinterface. Note that all data objects in spikeinterface are 'lazy' and do not actually contain the data, so these objects can be easily stored in classes.
`data_classes/base.py`  and `data_classes/preprocessing.py`: These data classes hold the raw data objects. At the moment, they don't do anything with them, but become more important when preprocessing is added.

Some notes:
1) 
The dataclass inheritence reduces boilerplate but is quite confusing, I am considering reverting back to normal classes just for readability. The way it works is the class attributes are placed at the start of the class. For sub-classing, only the class attributes of the subclass are placed at the start of the class. Then, when classing the class you must input all required arguments (first the superclasses, then the subclasses).

Because this only works for attributes, any logic must be put in a `__post_init__` function.

2) 
The idea is that the user will have data stored in SWCBlueprint style. This consists of subject, sessions, ephys level. Then, typically within each sessions there are different runs. The user must indicate the sessions and runs to load in the `sessions_and_runs` dictionary (e.g. `examples/load_data.py`). Then, the filesystem is traversed based on these names to load the specified sessions and runs into the `preprocessingData` class. This is a userdict (basically a dictionary with additional functions) in which the loaded data objects are stored in this dictionary. Later (future PR) these objects are operated on during preprocessing.

3)
Currently the test data is at:
`"/ceph/neuroinformatics/neuroinformatics/scratch/jziminski/ephys/test_data/steve_multi_run/1119617/time-short-multises"`

Please let me know if you have any trouble accessing it! Currently I am running off a mounted drive but let me know if you have trouble on macos with the path and I can update the usage instructions for the cluster (sorry these are currently unfinished).

### CODE REVIEW

File: spikewrap/pipeline/load_data.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    on the data filepaths.
+
+    TODO
+    ----
+    Figure out the format from the data itself, instead of passing as argument.
+    Do this when adding the next supported format.
+    """
+    empty_data_class = PreprocessingData(Path(base_path), sub_name, sessions_and_runs)
+
+    if data_format == "spikeglx":
```
Comment (lauraporta) [inline spikewrap/pipeline/load_data.py:58]: How many different loading there will be?
If they will be just two case we could have switch like:
```python
if data_format == "spikeglx":
    loaded_data = _load_spikeglx_data(empty_data_class)
elif data_format == "something_else":
    loaded_data = _load_something_else_data(empty_data_class)
else:
    raise RuntimeError("`data_format` not recognised.")
```

File: spikewrap/pipeline/load_data.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            se.read_spikeglx(
+                folder_path=run_path,
+                stream_id="imec0.ap",
+                all_annotations=True,
+                load_sync_channel=sync,
+            )
+            for sync in [True, False]
+        ]
+        preprocess_data[ses_name][run_name]["0-raw"] = without_sync
+        preprocess_data.sync[ses_name][run_name] = with_sync
```
Comment (lauraporta) [inline spikewrap/pipeline/load_data.py:91]: Here I see that the dict is very handy when you want to fill it in in iterations. 
However I am still not fully convinced about it, I don't know if it will be hard for the user to handle it

File: spikewrap/examples/load_data.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        "run-002_made_up_g0",
+    ],
+    "ses-002": [
+        "run-001_1119617_pretest1_shank12_g0",
+    ],
+    "ses-003": [
+        "run-002_1119617_pretest1_shank12_g0",
+    ],
+}
+
```
Comment (lauraporta) [inline spikewrap/examples/load_data.py:23]: How many dataset do you think a person will like to load simultaneously?
I think it is cool to save and read files maintaining the specs we have decided upon, however I think they should not be the keys of the data dict, because they are long, hard to read and typos are hard to spot, potentially leading to further confusions.
Previously I made simple classes with attributes `ses` `subjects` to hold the data information. But this does not have to be the case. We can discuss about it if you like to change this implementation.

File: spikewrap/pipeline/load_data.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        utils.message_user(f"Raw session data was loaded from {run_path}")
```
Comment (lauraporta) [inline spikewrap/pipeline/load_data.py:93]: What about using `logging.info(your_message`?

File: spikewrap/data_classes/base.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        ordered_ses_names = list(
+            chain(*[[ses] * len(runs) for ses, runs in self.sessions_and_runs.items()])
+        )
+        ordered_run_names = list(
+            chain(*[runs for runs in self.sessions_and_runs.values()])
+        )
+
+        return list(zip(ordered_ses_names, ordered_run_names))
+
+    def _validate_inputs(
```
Comment (lauraporta) [inline spikewrap/data_classes/base.py:52]: I like this set of `assert` statements, seems like a clean way to hold all the validations together.
Do you think you will define separate methods to get the folders? For instance will the `get_sub_path` change depending on the implementation of `load_data()`?
In case yes it makes sense to inject the `Callable`s

File: spikewrap/data_classes/base.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+from collections import UserDict
+from collections.abc import ItemsView, KeysView, ValuesView
+from dataclasses import dataclass
+from itertools import chain
+from pathlib import Path
+from typing import Callable, Dict, List, Literal
+
+
+@dataclass
+class BaseUserDict(UserDict):
```
Comment (lauraporta) [inline spikewrap/data_classes/base.py:11]: Will `BaseUserDict` be inherited by another class different than `PreprocessingData`? I am not sure on why you need two classes. 
Then also `load_data()` adds another layer of abstraction, actually finalising the datastructure by filling in the `.data` attribute. Maybe this could be simplified.

File: spikewrap/data_classes/base.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    def check_run_names_are_formatted_as_list(self) -> None:
+        """"""
+        for key, value in self.sessions_and_runs.items():
+            if not isinstance(value, List):
+                assert isinstance(
+                    value, str
+                ), "Run names must be string or list of strings"
+                self.sessions_and_runs[key] = [value]
+
+    def preprocessing_sessions_and_runs(self):  # TODO: type hint
```
Comment (lauraporta) [inline spikewrap/data_classes/base.py:41]: My inderstanding here is that you want to be sure that the keys in the nested dict structure (`data`) will be ordered. Is that correct?

File: spikewrap/data_classes/preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+@dataclass
+class PreprocessingData(BaseUserDict):
+    """
+    Dictionary to store SpikeInterface preprocessing recordings.
+
+    Details on the preprocessing steps are held in the dictionary keys e.g.
+    e.g. 0-raw, 1-raw-bandpass_filter, 2-raw_bandpass_filter-common_average
+    and recording objects are held in the value. These are generated
+    by the `pipeline.preprocess.run_preprocessing()` function.
```
Comment (lauraporta) [inline spikewrap/data_classes/preprocessing.py:19]: I guess the method `pipeline.preprocess.run_preprocessing()` is not part of this PR

File: spikewrap/data_classes/preprocessing.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        self._validate_rawdata_inputs()
+
+        self.sync: Dict = {}
+
+        for ses_name, run_name in self.preprocessing_sessions_and_runs():
+            utils.update(self.data, ses_name, run_name, {"0-raw": None})
+            utils.update(self.sync, ses_name, run_name, None)
```
Comment (lauraporta) [inline spikewrap/data_classes/preprocessing.py:49]: What is the use of `self.sync`?

**Review by lauraporta — COMMENTED** [review]
Interesting work, Joe!
I left some comments on parts of the code that are not clear to me. Open to discuss them when you have time!

File: spikewrap/data_classes/preprocessing.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    """
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        self._validate_rawdata_inputs()
+
+        self.sync: Dict = {}
+
+        for ses_name, run_name in self.preprocessing_sessions_and_runs():
+            utils.update(self.data, ses_name, run_name, {"0-raw": None})
```
[the remark this answers is not in the document]
Comment (JoeZiminski) [reply spikewrap/data_classes/preprocessing.py:48]: The `update` function has been fixed in #126 and further discussion on the dictionary key issue in #123

File: spikewrap/pipeline/load_data.py (hunk 2 above)
Comment (JoeZiminski) [reply spikewrap/pipeline/load_data.py:91]: In the end because this is not user-facing, and later on the values that hold recordings must be concatenated, we discussed leaving as a nested dictionary for now but to think about dict keys #123.

File: spikewrap/examples/load_data.py (hunk 3 above)
Comment (JoeZiminski) [reply spikewrap/examples/load_data.py:23]: We discussed this issue (dict keys #123) and general validation across projects that it would be nice to align (#124).

File: spikewrap/pipeline/load_data.py (hunk 4 above)
Comment (JoeZiminski) [reply spikewrap/pipeline/load_data.py:93]: This is a good point, for now the logging is quite complex because it needs to be propagated from SpikeInterface. For now I will keep the `utils.message_user()` as it provides some flexibility down the line when it come to implemented the GUI. It is not in this PR but there is some distinction between `spikewrap` controlled print-messages and the print messages `spikeinterface` outputs which are all funneleld to logs. But this definately warrants further discussion on an upcoming logging PR.

File: spikewrap/data_classes/base.py (hunk 6 above)
Comment (JoeZiminski) [reply spikewrap/data_classes/base.py:11]: Later on there are `SortingData` class and it's subclasses that inhereit from this. However, we discussed the factory -method approach for `load_data` also. I will make a issue on this as OpenEphys support needs to be added before the first version release and there is somone interested now with test data.

File: spikewrap/data_classes/base.py (hunk 7 above)
Comment (JoeZiminski) [reply spikewrap/data_classes/base.py:41]: This unecessarily complex function has been refactored in #121 (it took me ages to figure out what it was actually doing 😅)

File: spikewrap/data_classes/preprocessing.py (hunk 8 above)
Comment (JoeZiminski) [reply spikewrap/data_classes/preprocessing.py:19]: This will be next PR 😅

File: spikewrap/data_classes/preprocessing.py (hunk 9 above)
Comment (JoeZiminski) [reply spikewrap/data_classes/preprocessing.py:49]: This is the synchonrisation channel that contains TTL pulse information. SI works in that the data used for sorting cannot contain the sync channel, however in `spikewrap` we want the sync channel as well. So we need to load it twice (once with once without sync). This could be made clearer though in the docs, maybe there is a cleaner way to write this.

File: spikewrap/data_classes/preprocessing.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (JoeZiminski) [reply spikewrap/data_classes/preprocessing.py:48]: and #127 addresses the confusing `UserDict`, `self.data` references.

File: spikewrap/pipeline/load_data.py (hunk 1 above)
Comment (JoeZiminski) [reply spikewrap/pipeline/load_data.py:58]: Great point we may support 5+ file formats so an important consideration, opened #128 and can consider further when implementing OpenEphys support.
